### PR TITLE
Prevent publishing of empty tests

### PR DESCRIPTION
**Pull request by martijn-tao**
Added the same check that is implemented on the test search on the deliveries page.

### CODE REVIEW

File: controller/Publish.php [hunk 1]
```diff
@@ -58,12 +58,21 @@ public function publish(){
             $testUri = $this->getRequestParameter('testUri');
             $classUri = $this->getRequestParameter('classUri');
             $test = $this->getResource($testUri);
+
+            $testService = \taoTests_models_classes_TestsService::singleton();
```
Comment (siarhei-baradzin) [inline controller/Publish.php]: Small notice: when I use old `OntologyClassService`s I prefer to retrieve it like:
```
$this->getServiceLocator()->get(\taoTests_models_classes_TestsService::class)
```
(knowing that `singleton()` call does the same).
It looks more uniform with other services retrieval and allows you to mock them in tests.

File: controller/Publish.php [hunk 2]
```diff
@@ -58,12 +58,21 @@ public function publish(){
...
+            $testItems = $testService->getTestItems($test);
+            if (empty($testItems)) {
+                throw new \RuntimeException(
```
Comment (boajer) [inline controller/Publish.php]: Can you use Simplified FQN ?

File: controller/Publish.php [hunk 3]
```diff
@@ -58,12 +58,21 @@ public function publish(){
...
+                );
+            }
+
             $deliveryClass = $this->getClass($classUri);
             /** @var DeliveryFactory $deliveryFactoryResources */
             $deliveryFactoryResources = $this->getServiceManager()->get(DeliveryFactory::SERVICE_ID);
             $initialProperties = $deliveryFactoryResources->getInitialPropertiesFromArray([OntologyRdfs::RDFS_LABEL => 'new delivery']);
             return $this->returnTaskJson(CompileDelivery::createTask($test, $deliveryClass, $initialProperties));
-        }catch(\Exception $e){
+        } catch(\Exception $e) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline controller/Publish.php]: FQN :)

File: views/js/controller/Publish/index.js [hunk 4]
```diff
@@ -95,6 +95,9 @@ define([
                 refreshTree(testUri);
             })
             .on('error', function(err){
+                if (err && err.errorMsg) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline views/js/controller/Publish/index.js]: please ask a FE Engineer to validate this  :)

Comment (martijn-tao) [reply views/js/controller/Publish/index.js]: > `index.js` change approved.

- @oatymart 

;)

Comment (oatymart) [reply views/js/controller/Publish/index.js]: Already approved ✔️

File: controller/Publish.php [hunk 5]
```diff
@@ -55,15 +58,26 @@ public function index(){
...
+            $parsedBody = $this->getPsrRequest()->getParsedBody();
+            $testUri = $parsedBody['testUri'] ?? null;
+            $classUri = $parsedBody['classUri'] ?? null;
             $test = $this->getResource($testUri);
+
+            /** @var taoTests_models_classes_TestsService $testService */
+            $testService = $this->getServiceLocator()->get(taoTests_models_classes_TestsService::class);
+            $testItems = $testService->getTestItems($test);
+            if (empty($testItems)) {
+                throw new RuntimeException(
```
Comment (julien-sebire) [inline controller/Publish.php]: With a little refactoring, it seems that only three things can throw an exception in the whole try:
- the two calls to getServiceLocator()->get(), which would mean a wiring problem and should crash the application
- the exception thrown here manually.
=> we can directly return the failure response.

**Pull request merged**